torch/_inductor/kernel/bmm.py

```diff
@@ -87,6 +88,31 @@ def bmm_grid(b, m, n, meta):
 
 @register_lowering(aten.bmm)
 def tuned_bmm(mat1, mat2, *, layout=None):
+    # Make the inputs of bmm contiguous
+    # because bmm cpu implementation does contiguous() if not
+    # this is to avoid additional copies in bmm
+    def do_bmm_input_contiguous(t: ir.TensorBox):
+        if (
+            isinstance(t.data, ir.View)
+            and isinstance(t.data.data, ir.PermuteView)
+            and t.data.data.dims == [0, 3, 1, 2]
+        ):
+            t = ir.Pointwise.create(
+                device=t.get_device(),
+                dtype=t.get_dtype(),
+                inner_fn=t.make_loader(),
+                ranges=t.get_size(),
+                origin_node=t.get_origin_node(),
+                traceback=t.get_traceback(),
+            )
+            t.realize()
+            t.freeze_layout()
```

Comment:
Several issues with this change:
1. It only handles a specific case where the tensor is in 4D and permuted with a particular order. Can we make it general? Basically, we want a particular order of the last two dims?
2. Related to 1, bmm can actually handle non-contiguous cases and also transposed cases for the last two dims. It only requires one of the dims to be contiguous while the other can have a stride larger than the size of the former dim. Is it too strict to always force contiguous here?
3. Perhaps we can call `require_stride_order` instead of implementing the `copy_input` again here?

Reply:
Thanks for your suggestions!

A new function called `is_pointwise_contiguous_or_transposed_after_perm` is added to deduce the layout of a pointwise from its readers, because it has not been realized yet.